origins_toc/origins_toc.module

```diff
@@ -0,0 +1,291 @@
+    '#title' => t('Title'),
+    '#description' => t('Displayed before the table of contents or blank for no title.'),
+    '#default_value' => $node->getThirdPartySetting('origins_toc', 'toc_title', 'Table of contents'),
+  ];
+
+  $fields = \Drupal::service('entity_field.manager')->getFieldDefinitions('node', $node_type);
+
+  foreach ($fields as $field) {
+    if ($field instanceof FieldConfig) {
+      $field_options[$field->getName()] = $field->label();
+    }
+  }
+
+  $form['toc']['toc_config']['toc_source_field'] = [
+    '#type' => 'select',
+    '#title' => t('Source field'),
+    '#options' => $field_options,
+    '#description' => t('Select the field to build the table of contents for.'),
+    '#default_value' => $node->getThirdPartySetting('origins_toc', 'toc_source_field', NULL),
+  ];
+
+  $form['toc']['toc_config']['toc_element'] = [
+    '#type' => 'select',
+    '#title' => t('Source element'),
+    '#options' => [
+      'h2' => 'H2 - Heading 2',
+      'h3' => 'H3 - Heading 3',
+      'h4' => 'H4 - Heading 4',
+    ],
+    '#description' => t('HTML element to generate the table of contents against.'),
+    '#default_value' => $node->getThirdPartySetting('origins_toc', 'toc_element', NULL),
+  ];
+
+  $form['toc']['toc_config']['toc_exclusions'] = [
+    '#type' => 'textfield',
+    '#title' => t('Exclusions'),
+    '#description' => t('Comma separated list of class names to exclude from the table of contents.'),
+    '#default_value' => $node->getThirdPartySetting('origins_toc', 'toc_exclusions', NULL),
+  ];
+
+  $form['#entity_builders'][] = 'origins_toc_form_node_type_form_builder';
+}
+
+/**
+ * Entity builder for the origins_toc_form_node_type_form_alter().
+ */
+function origins_toc_form_node_type_form_builder($entity_type, $type, &$form, $form_state) {
+  $type->setThirdPartySetting('origins_toc', 'toc_enable', $form_state->getValue('toc_enable'));
+  $type->setThirdPartySetting('origins_toc', 'toc_title', $form_state->getValue('toc_title'));
+  $type->setThirdPartySetting('origins_toc', 'toc_source_field', $form_state->getValue('toc_source_field'));
+  $type->setThirdPartySetting('origins_toc', 'toc_element', $form_state->getValue('toc_element'));
+  $type->setThirdPartySetting('origins_toc', 'toc_exclusions', $form_state->getValue('toc_exclusions'));
+}
+
+/**
+ * Implements hook_entity_presave().
+ */
+function origins_toc_entity_presave(EntityInterface $entity) {
+  if ($entity instanceof NodeType) {
+    $toc_enabled = $entity->getThirdPartySetting('origins_toc', 'toc_enable', FALSE);
+    if ($toc_enabled) {
+      // If 'toc enable' doesn't exist for this node, create it.
+      $field = FieldConfig::loadByName('node', $entity->id(), 'field_toc_enable');
+      if (!$field) {
+        // Create the enabled field for this entity.
+        $field_storage = FieldStorageConfig::loadByName('node', 'field_toc_enable');
+
+        // If the field storage config doesn't exists, warn and abort.
```

Comment:
Might it be better to create the storage programmatically rather than error in this case ?
(I did this in the origins_workflow module, in AuditSettingsForm.php)

Reply:
Yeah, I've kinda hmm'd and haaa'd about this and stuck with the yaml file just to get it up and running. And yeah, programmatically is a safer way of doing it.

Reply:
Updated it to create in code 👍